### PR TITLE
Add a hack to remove output from the Razor generator if suppressed

### PR DESCRIPTION
**Pull request by jasonmalinowski**
The Razor generator is controlled by a flag that lives in an .editorconfig file; in the IDE we generally don't run the generator
and instead use the design-time files added through the legacy IDynamicFileInfo API. When we're doing Hot Reload we then
remove those legacy files and remove the .editorconfig file that is supposed to disable the generator; for the Hot Reload pass we then are running the generator. This is done in the CompileTimeSolutionProvider.

https://github.com/dotnet/sdk/pull/24928 introduced an issue where even though the Razor generator is being told to not run, it still runs anyways. As a tactical fix rather than reverting that PR, for Visual Studio 17.3 Preview 2 we are going to do a hack here which is to rip out generated files. In discussion, simply reverting the SDK change was problematic because it would reintroduce some notable performance issues that it was intended to fix in the first place. Fixing the generator directly is also possible (and we'll be doing a follow-up PR to try doing that), but is also a fairly risky change given the amount of testing that was required to test the PR in the first place. This we think is the least-risky option we have, even though it itself is touching a fairly sensitive area of our codebase.

The intent is to remove this hack by Preview 3 with a properly tested fix in the generator.

This is a tactical fix for https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1545938

### CODE REVIEW

**Review by davidwengier — APPROVED** [review]

**Comment by davidwengier** [comment]
Random question @jasonmalinowski: When Ryan and I were debugging this we could see the source generated documents in the analyzers tree in Solution Explorer. It didn't click at the time, but that was probably the sign that things were wrong, because that is unexpected, correct? And if so, does this fix remove them again?

**Comment by jasonmalinowski** [comment]
@davidwengier Yes, "unfortunately" -- that would have been a hint indeed that something was wrong/changed from before.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs [hunk 1]
```diff
@@ -1078,6 +1103,35 @@ public async ValueTask<ImmutableArray<Diagnostic>> GetSourceGeneratorDiagnostics
...
+            // run, it still runs anyways. As a tactical fix rather than reverting that PR, for Visual Studio 17.3 Preview 2 we are going
+            // to do a hack here which is to rip out generated files.
+
+            private bool IsGeneratorRunResultToIgnore(GeneratorRunResult result)
+            {
+                var globalOptions = this.ProjectState.AnalyzerOptions.AnalyzerConfigOptionsProvider.GlobalOptions;
+
+                // This matches the implementation in https://github.com/chsienki/sdk/blob/4696442a24e3972417fb9f81f182420df0add107/src/RazorSdk/SourceGenerators/RazorSourceGenerator.RazorProviders.cs#L27-L28
+                var suppressGenerator = globalOptions.TryGetValue("build_property.SuppressRazorSourceGenerator", out var option) &&
+                                                               option == "true";
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs:1122]:
```suggestion
                var suppressGenerator = globalOptions.TryGetValue("build_property.SuppressRazorSourceGenerator", out var option) && option == "true";
```

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by davidwengier** [comment]
> @davidwengier Yes, "unfortunately" -- that would have been a hint indeed that something was wrong/changed from before.

Would adding a manual check for the presence of files under the Analyzers node be a good early warning sign, to avoid the long investigation through multiple teams to track this down? Or is it no better a signal than "the whole document has red squiggles" which got us here?

**Comment by jasonmalinowski** [comment]
> Would adding a manual check for the presence of files under the Analyzers node be a good early warning sign, to avoid the long investigation through multiple teams to track this down?

@davidwengier Yes, potentially, but the red squiggles all about ambiguous types was also a pretty good giveaway if you know how this system is supposed to work too. We're also updating tests for the generator properly in https://github.com/dotnet/sdk/pull/25833 so long investigations don't happen either. :smile:

**Pull request merged**